sqsjobs/item.go

```diff
@@ -161,36 +169,48 @@ func (i *Item) Nack() error {
 		return nil
 	}
 
-	// requeue message
-	err := i.Options.requeueFn(context.Background(), i)
-	if err != nil {
-		return err
-	}
+	if !i.Options.RetainFailedJobs {
+		// requeue as new message
+		err := i.Options.requeueFn(context.Background(), i)
+		if err != nil {
+			return err
+		}
+		// Delete original message
+		_, err = i.Options.client.DeleteMessage(context.Background(), &sqs.DeleteMessageInput{
+			QueueUrl:      i.Options.queue,
+			ReceiptHandle: i.Options.receiptHandler,
+		})
 
-	_, err = i.Options.client.DeleteMessage(context.Background(), &sqs.DeleteMessageInput{
-		QueueUrl:      i.Options.queue,
-		ReceiptHandle: i.Options.receiptHandler,
-	})
+		if err != nil {
+			return err
+		}
+	} else if i.Options.ErrorVisibilityTimeout > 0 {
+		// If error visibility is defined change the visibility timeout of the job that failed
+		_, err := i.Options.client.ChangeMessageVisibility(context.Background(), &sqs.ChangeMessageVisibilityInput{
+			QueueUrl:          i.Options.queue,
+			ReceiptHandle:     i.Options.receiptHandler,
+			VisibilityTimeout: i.Options.ErrorVisibilityTimeout,
+		})
 
-	if err != nil {
-		return err
-	}
+		if err != nil {
+			var notInFlight *types.MessageNotInflight
+			if goErrors.As(err, &notInFlight) {
+				// I would like to log info/warning here, but not sure how to get the correct logger
+				log.Println("MessageNotInFlight; ignoring ChangeMessageVisibility")
```

Comment:
Do not use `log` here. You may return the error and do type assertion (goErrors -> stderr) in the caller.

Reply:
I'd suggest here to just ignore that, without log. Also, this specific type of error should be tested at least by hands. AWS error are deeeply nested, so, NoInFlight error might be hidden by 2-3 additional error types.

Reply:
Is the caller in `listener.go`? Not sure where to put this exactly.

Reply:
Ah, caller, here is the JOBS plugin. Nevermind. We just need to check for this type of error, put a comment on why we're skipping error, and return nil. There are no logs in the item and honestly speaking, I would rather not overload the Options structure with a logger field.
I just see no reason to spam with that log message. WDYT?

Reply:
This can happen if something takes so long to process (before it errors) that the job becomes visible again, but has not yet been consumed. In that case, changing the visibility should be ignored as the message is already visible. I don't know if we need to log this, but I thought it would be nice for a debug/info at least. Obviously, logging of this kind is never strictly necessary. If you think it breaks a pattern or introduces unnecessary imports/declarations, we can drop it.

Reply:
> I'd suggest here to just ignore that, without log. Also, this specific type of error should be tested at least by hands. AWS error are deeeply nested, so, NoInFlight error might be hidden by 2-3 additional error types.

According to docs for this method, it should be returned directly from this method: https://pkg.go.dev/github.com/aws/aws-sdk-go/service/sqs#SQS.ChangeMessageVisibility
```
Returned Error Types:

MessageNotInflight The specified message isn't in flight.

ReceiptHandleIsInvalid The specified receipt handle isn't valid.
...
```

Reply:
Please rename goErrors to stderr